Create features.md

bamurtaugh opened the pull request:
Add the content from https://github.com/devcontainers/spec/blob/main/proposals/devcontainer-features.md. 

One to-do is to make the headers into links (i.e. `## <a href="#contribution-approaches" name="contribution-approaches" class="anchor"> Contribution approaches </a>`), but that could also be done in a subsequent PR.

bamurtaugh committed:
Create features.md

bamurtaugh committed:
Update contributing.md

Chuxel commented:
Yeah I think this would be reasonable to add at this point along with a nav! We can then link to it from the devcontainer.json reference and the main spec. A few thoughts:
1. Maybe we should add `[Proposal]` to the heading with a brief statement saying something like "*This section provides information on a currently active proposal. See https://github.com/devcontainers/spec/issues/61 for input and links to other proposed improvements.* "?
1. It looks like this references the distribution proposal directly. We may want to pull that file in at the same time or adjust the link to point to the markdown.
1. We could reference the features section from the main spec I think just by adding a heading under "Other options". E.g., "Dev container 'features' are self-contained, shareable units of installation code and development container configuration. They are applied to container images as a secondary build step and can affect a number of dev container configuration settings. See <link> for more details."
1. In the "features" property in the devcontainer.json reference, we can reference the "devcontainer.json properties" section of this doc for more info at this point.

Commit:
Add features distribution, wording updates

Commit:
Add features.md headers

Commit:
Add headers

Commit:
Cross-linking

bamurtaugh commented:
Thanks for the great feedback, @Chuxel!

I believe I addressed the 4 points above:
1. Added the word 'Proposal' and a sentence of context to the features doc
2. Pulled in the features distribution doc directly, also using 'Proposal' and the context sentence above
3. Added a features section under "Other Options" in the main spec, linking to the new features.md
4. Added a link to features.md to the features section of the devcontainer.json reference

I also made all the headers in the features and features distribution docs into links for easier sharing and navigation, consistent with the other pages of the site.

Please let me know if the changes are in-line with what you were thinking.

Chuxel commented on _implementors/features.md:
```diff
@@ -0,0 +1,267 @@
+
+> Note: This section provides information on a currently active proposal. See the [Features proposal in the spec repo](https://github.com/devcontainers/spec/issues/61) for input and links to other proposed improvements.
+
+Dev container 'features' are self-contained, shareable units of installation code and development container configuration.
```
```suggestion
Development container "Features" are self-contained, shareable units of installation code and development container configuration. The name comes from the idea that referencing one of them allows you to quickly and easily add more tooling, runtime, or library "features" into your development container for you or your collaborators to use.
```

Possible tweak to the description - we could back port this to the main spec too.  Gives a little context on why we called them "Features".

Chuxel commented on _implementors/json_reference.md:
```diff
@@ -54,7 +54,7 @@ The focus of `devcontainer.json` is to describe how to enrich a container for th
 | `userEnvProbe` | enum | Indicates the type of shell to use to "probe" for user environment variables to include in `devcontainer.json` supporting services' / tools' processes: `none`, `interactiveShell`, `loginShell`, or `loginInteractiveShell` (default). The specific shell used is based on the default shell for the user (typically bash). For example, bash interactive shells will typically incl ...
 | `overrideCommand` | boolean | Tells `devcontainer.json` supporting services / tools whether they should run `/bin/sh -c "while sleep 1000; do :; done"` when starting the container instead of the container's default command (since the container can shut down if the default command fails). Set to `false` if the default command must run for the container to function properly. Defaults to `true` fo ...
-| `features` (currently in preview) | object | An object of dev container features and related options to be added into your primary container. The specific options  that are available varies by feature, so see its documentation for additional details. For example: <br />`"features": {"github-cli": "latest"}` <br /><br /> ⚠️ Currently in preview. |
+| `features` (currently in preview) | object | An object of dev container features and related options to be added into your primary container. The specific options  that are available varies by feature, so see [its documentation](./features.md#a-hrefdevcontainer-json-properties-namedevcontainer-json-properties-classanchor-devcontainerjson-properties-a) for additional details. For example: <br /> ...
```
```suggestion
| `features` (proposed) | object | An object of dev container features and related options to be added into your primary container. The specific options  that are available varies by feature, so see [its documentation](./features.md#a-hrefdevcontainer-json-properties-namedevcontainer-json-properties-classanchor-devcontainerjson-properties-a) for additional details. For example: <br />`"features": {"github-cli": "latest"}` <br /><br /> ⚠️ Support is currently in preview as the [spec](./features.md) is not yet final. |

Maybe we specifically refer to this as proposed to tie the two together?
```

Chuxel commented on _implementors/spec.md:
```diff
@@ -67,6 +67,10 @@ In addition to the configuration options explained above, there are other settin
 
+## <a href="#features-preview" name="features-preview" class="anchor"> Features (preview) </a>
+
+Dev container 'Features' are self-contained, shareable units of installation code and development container configuration. They are applied to container images as a secondary build step and can affect a number of dev container configuration settings. See the [features documentation](./features.md) for more details.
```
```suggestion
Development container "Features" are self-contained, shareable units of installation code and development container configuration. The name comes from the idea that referencing one of them allows you to quickly and easily add more tooling, runtime, or library "features" into your development container for you or your collaborators to use.

They are applied to container images as a secondary build step and can affect a number of dev container configuration settings. See the [features documentation](./features.md) for more details.
```

Same tweak suggestion as above.

Chuxel reviewed (approved):
A few suggested tweaks.

bamurtaugh committed:
Update _implementors/features.md

Co-authored-by: Chuck Lantz <clantz@microsoft.com>

bamurtaugh committed:
Update _implementors/json_reference.md

Co-authored-by: Chuck Lantz <clantz@microsoft.com>

bamurtaugh committed:
Update _implementors/spec.md

Co-authored-by: Chuck Lantz <clantz@microsoft.com>

bamurtaugh commented:
Thanks for the suggestions! Opened https://github.com/devcontainers/spec/pull/69 with the Features wording suggestion you mentioned.

I'll merge these changes in as it seems this is a good baseline, and we can continue with other changes folks might have in mind in follow-up PRs.
